Update README.md (#4)
redhat-et/optimal-stopping-point-tool: README.md

# Optimal Stopping Point

> The aim of this tool is to predict an optimal stopping point for CI tests based on their test duration (runtimes). We perform an initial data analysis as well as feature engineering on the CI test data. Furthermore, we also calculate the optimal stopping point by identifying the distribution of the test duration values for different CI tests and comparing the distributions between the passing and failing tests.

Reviewer: What feature engineering is done on the raw data? I thought we were collecting and training on run duration.

Author: Yeah, we are collecting and training on test duration, but we also try to evaluate various distributions to find the best fit and compare the p values before finalizing what model we want to use. All of this would be a part of this feature engineering process. If you think this name isn't suitable we can call it "Fit distribution" or "Finding best distribution" also.

Reviewer: I would not call determining the best fit distribution to be feature engineering. I thought feature engineering was more along the lines of transforming the raw inputs in some way before applying them  to the model so it can learn the underlying task better.

Maybe you could say that you are performing "automated model selection"?

Reviewer 2: I like something on the lines of "Finding best distribution"

Reviewer 2: Are we training a model? https://github.com/redhat-et/optimal-stopping-point-tool/pull/4#discussion_r1094935509

## Data Collection

> We fetch the CI test data from github and further store this to s3 storage. The main requirement here is to get test duration values. We can visualize the distribution of the test durations and based on the distribution type identified, we can find a point after which the test has a higher probability of failing.

Reviewer 2: ```suggestion
We collect historical test data mainly the test duration values from running workflows on Github and store the data on s3 storage.
```

> We fetch the CI test data from github and further store this to s3 storage. The main requirement here is to get test duration values. We can visualize the distribution of the test_duration metric across various testgrid dashboards and jobs. Based on the distribution type identified, we can find a point after which the test has a higher probability of failing.

Reviewer: Is testgrid relevant for this project? I thought we were ignoring that and working directly with the github data.

Author: I fixed it. My bad.

## Feature Engineering

> After fetching the data, we approximate the distributions of the test duration value and also check its goodness of fit for different tests. Based on the type of distribution identified, we can calculate the probability of the test failing.

Reviewer 2: ```suggestion
We then visualize the distribution of the test durations and based on the distribution type identified, we can find a point after which the test has a higher probability of failing. We approximate the distributions of the test duration value and also check its goodness of fit for different tests. Based on the type of distribution identified, we calculate the probability of the test failing.
```

Reviewer 2: "check its goodness of fit for different tests" - should this be goodness of fit for different distributions? This paragraph is a little confusing, can we rephrase?

## Model Training

After performing initial data analysis and calulating the probability to fail, we predict the optimal stopping for a given test based on their test duration(runtimes). We find the best distribution(s) for the given test, and find an optimal stopping point for the test by finding the point where the probability of the test failing is greater than the probabilty of the test passing.

> Once this model is trained, it will use the Optimal stopping point value to further stop long running tests from consuming the resources after a certain point. This can help the engineering teams and managers better allocate their resources.

Reviewer 2: ```suggestion
Once we have calculated optimal stopping point values for tests, this can be used to stop long running tests from consuming resources after a certain point. This can help the engineering teams and managers better allocate their resources.
```

Is there a model training step here? Do we have a model or are we essentially calculating intersections between distribution curves? Should we rename the step to something better suited?

Author: Yes, we basically use statistical models to predict the distributions for the test durations and get the intersections. Maybe we should rename this. Maybe predict OSP, What do you think?

Author: Updated it to "Predicting the OSP" lmk wdyt?

Author: Awesome, updated.
